lnst/Tests/PktGen.py

```diff
@@ -0,0 +1,157 @@
+
+from lnst.Common.Utils import kmod_in_use
+from lnst.Common.IpAddress import ipaddress, Ip4Address
+from lnst.Tests.BaseTestModule import BaseTestModule, TestModuleError
+from lnst.Common.Parameters import IntParam, IpParam, StrParam, ListParam
+
+
+class PktGenResultsParser:
+    def __init__(self, devs: list[str], res: dict) -> None:
+        self._devs = devs
+        self._res = res
+
+    def _read_dev_outputs(self) -> Iterator[tuple[str, str]]:
+        for device in self._devs:
+            output = ""
+            with open(f"/proc/net/pktgen/{device}", "r") as f:
+                output = f.readlines()
+            yield device, "\n".join(output)
+
+    def _split_output(self, output: str) -> tuple:
+        match = re.search(r"Params:(.+)Current:(.+)Result:\s(\w+)", output, re.DOTALL)
+        if not match:
+            raise TestModuleError("Could not parse pktgen devide output")
+
+        return match.groups()
+
+    def _parse_values(self, params) -> dict[str, str]:
+        values = {}
+
+        for key, value in re.findall(r"(\w+):?\s(\S+)", params, re.MULTILINE):
+            values[key.lower()] = value
+
+        return values
+
+    def parse_dev_outputs(self):
+        for device, output in self._read_dev_outputs():
+            params, current, state = self._split_output(output)
+
+            self._res[device] = {
+                "params": self._parse_values(params),
+                "current": self._parse_values(current),
+                "state": state,
+            }
+
+
+class PktGen(BaseTestModule):
+    """
+    In the scope of this module, the physical interface is refered as `interface`.
+    Pktgen device (interface@anything) is refered as device.
+
+    Inspired by https://github.com/torvalds/linux/blob/master/samples/pktgen/pktgen_sample03_burst_single_flow.sh
+    """
+
+    cpus = ListParam(type=IntParam())  # each CPU is 1 generator
+
+    pgctrl = StrParam(mandatory=False, default="/proc/net/pktgen/pgctrl")
+
+    src_mac = StrParam(mandatory=False, default="90:e2:ba:ff:ff:ff")
+    dst_mac = StrParam()
+
+    src_ip = IpParam(
+        mandatory=False, default=ipaddress("192.168.1.1")
+    )  # IP doesn't really matter
+    dst_ip = IpParam()
+
+    interface = StrParam()
+
+    count = IntParam(mandatory=False, default=0)  # 0 = no upper limit
+    size = IntParam(mandatory=False, default=60)  # 4 bytes are added for CRC by NIC
+    frags = IntParam(mandatory=False, default=1)
+    burst = IntParam(mandatory=False, default=8)
+
+    duration = IntParam(default=60)
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+        if not kmod_in_use("pktgen"):
+            raise TestModuleError("pktgen module is not loaded")
+
+        self._devices = []
+        self._ip6 = True
+        if isinstance(self.params.dst_ip, Ip4Address):
+            self._ip6 = False
+
+        self._res_data = {}
+        self._output_parser = PktGenResultsParser(self._devices, self._res_data)
```

Comment:
similar issue to the pktgen module - this shouldn't be instantiated in the `__init__` method but at the start of the `run()` method which is only called on the Agent side.

Reply:
Thanks